File: C++/Relationship.cpp
// Notification system — Dependency Injection via composition
#include <iostream>
#include <string>
using namespace std;

class INotification {
public:
    virtual void sendNotification(const string& message) = 0;
    virtual ~INotification() {}
};

class SMSNotification : public INotification {
public:
    void sendNotification(const string& message) override {
        cout << "SMS: " << message << "\n";
    }
};

class EmailNotification : public INotification {
public:
    void sendNotification(const string& message) override {
        cout << "Email: " << message << "\n";
    }
};

class WhatsAppNotification : public INotification {
public:
    void sendNotification(const string& message) override {
        cout << "WhatsApp: " << message << "\n";
    }
};

class SlackNotification : public INotification {
public:
    void sendNotification(const string& message) override {
        cout << "Slack: " << message << "\n";
    }
};

class NotificationClient {
    INotification* notification;
public:
    NotificationClient() {
        notification = new SMSNotification();
    }

    void setNotification(INotification* n) {
        notification = n;
    }

    void sendNotification(const string& message) {
        notification->sendNotification(message);
    }

    ~NotificationClient() { delete notification; }
};

int main() {
    NotificationClient client;
    client.sendNotification("Hello via SMS");

    SlackNotification slack;
    client.setNotification(&slack);
    client.sendNotification("Hello via Slack");

    return 0;
}

File: Python/Relationship.py
# Notification system — Dependency Injection via composition
from abc import ABC, abstractmethod


class INotification(ABC):
    @abstractmethod
    def send_notification(self, message: str): pass


class SMSNotification(INotification):
    def send_notification(self, message: str):
        print(f"SMS: {message}")


class EmailNotification(INotification):
    def send_notification(self, message: str):
        print(f"Email: {message}")


class WhatsAppNotification(INotification):
    def send_notification(self, message: str):
        print(f"WhatsApp: {message}")


class SlackNotification(INotification):
    def send_notification(self, message: str):
        print(f"Slack: {message}")


class NotificationClient:
    def __init__(self):
        self.notification: INotification = SMSNotification()

    def set_notification(self, notification: INotification):
        self.notification = notification

    def send_notification(self, message: str):
        self.notification.send_notification(message)


if __name__ == "__main__":
    client = NotificationClient()
    client.send_notification("Hello via SMS")

    client.set_notification(SlackNotification())
    client.send_notification("Hello via Slack")
